### PR TITLE
add memory read/write abilities to python windows meterpreter

### PR DESCRIPTION
**Pull request by NtAlexio2**
This PR adds two memory functions to python meterpreter on windows platform:
- stdapi_sys_process_memory_read
- stdapi_sys_process_memory_write

Tested with:
- Python 3.7.2
- Python 2.7.15

Test output:
```
meterpreter > sysinfo 
Computer        : DESKTOP-5HNNBGG
OS              : Windows 10 (Build 22000)
Architecture    : x64
System Language : en_GB
Meterpreter     : python/windows
meterpreter > pry
[*] Starting Pry shell...
[*] You are in the "client" (session) object

[1] pry(#<Msf::Sessions::Meterpreter_Python_Python>)> proc = sys.process.open(6152)
=> #<#<Class:0x00007f098a1dc550>:0x00007f098a7c6f88
 @aliases=
  {"image"=>#<Rex::Post::Meterpreter::Extensions::Stdapi::Sys::ProcessSubsystem::Image:0x00007f098a7c6e20 @process=#<#<Class:0x00007f098a1dc550>:0x00007f098a7c6f88 ...>>,
   "io"=>#<Rex::Post::Meterpreter::Extensions::Stdapi::Sys::ProcessSubsystem::IO:0x00007f098a7c6dd0 @process=#<#<Class:0x00007f098a1dc550>:0x00007f098a7c6f88 ...>>,
   "memory"=>#<Rex::Post::Meterpreter::Extensions::Stdapi::Sys::ProcessSubsystem::Memory:0x00007f098a7c6d58 @process=#<#<Class:0x00007f098a1dc550>:0x00007f098a7c6f88 ...>>,
   "thread"=>#<Rex::Post::Meterpreter::Extensions::Stdapi::Sys::ProcessSubsystem::Thread:0x00007f098a7c6d08 @process=#<#<Class:0x00007f098a1dc550>:0x00007f098a7c6f88 ...>>},
 @channel=nil,
 @client=#<Session:meterpreter 10.0.0.2:49694 (10.0.0.2) "DESKTOP-5HNNBGG\John @ DESKTOP-5HNNBGG">,
 @handle=916,
 @pid=6152>
[2] pry(#<Msf::Sessions::Meterpreter_Python_Python>)> proc.memory.write(addr, "Hello world!")
Rex::Post::Meterpreter::RequestError: stdapi_sys_process_memory_write: Operation failed: Windows error: 998
from /home/kali/git/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/sys/process_subsystem/memory.rb:176:in `write'
[3] pry(#<Msf::Sessions::Meterpreter_Python_Python>)> addr = proc.memory.allocate(2048)
=> 17367040
[4] pry(#<Msf::Sessions::Meterpreter_Python_Python>)> proc.memory.write(addr, "Hello world!")
=> 12
[5] pry(#<Msf::Sessions::Meterpreter_Python_Python>)> proc.memory.read(addr, 12)
=> "Hello world!"
[6] pry(#<Msf::Sessions::Meterpreter_Python_Python>)>
```

### CODE REVIEW

File: python/meterpreter/ext_server_stdapi.py [hunk 1]
```diff
@@ -1663,6 +1663,44 @@ def stdapi_sys_process_memory_unlock(request, response):
...
+    readed_data = buffer.raw[:bytes_read.value]
+    response += tlv_pack(TLV_TYPE_PROCESS_MEMORY, readed_data)
+    return ERROR_SUCCESS, response
+
+@register_function_if(has_windll)
+def stdapi_sys_process_memory_write(request, response):
+    ERROR_PARTIAL_COPY = 229
+    handle = packet_get_tlv(request, TLV_TYPE_HANDLE).get('value', 0)
+    base = packet_get_tlv(request, TLV_TYPE_BASE_ADDRESS).get('value', 0)
+    data = packet_get_tlv(request, TLV_TYPE_PROCESS_MEMORY).get('value', 0)
```
Comment (smcintyre-r7) [inline python/meterpreter/ext_server_stdapi.py:1691]: This doesn't seem like a good idea. If `TLV_TYPE_PROCESS_MEMORY` isn't present, `data` will be set to 0 when it should be bytes.

`data` should always be either `None` or a `bytes` object (in Python 3 and `str` in Python 2 of course).

Really though, if any of these three parameters are not present, it should fail with `ERROR_INVALID_PARAMETER` as the Windows one does in [memory.c](https://github.com/rapid7/metasploit-payloads/blob/6bac251842e0140b78e6bd7e23e56688b83f34cb/c/meterpreter/source/extensions/stdapi/server/sys/process/memory.c#L160-L167).

File: python/meterpreter/ext_server_stdapi.py [hunk 2]
```diff
@@ -1663,6 +1663,44 @@ def stdapi_sys_process_memory_unlock(request, response):
         return error_result_windows(), response
     return ERROR_SUCCESS, response
 
+@register_function_if(has_windll)
+def stdapi_sys_process_memory_read(request, response):
+    ERROR_PARTIAL_COPY = 229
+    handle = packet_get_tlv(request, TLV_TYPE_HANDLE).get('value', 0)
+    base = packet_get_tlv(request, TLV_TYPE_BASE_ADDRESS).get('value', 0)
+    size = packet_get_tlv(request, TLV_TYPE_LENGTH).get('value', 0)
```
Comment (smcintyre-r7) [inline python/meterpreter/ext_server_stdapi.py:1671]: This should fail with `ERROR_INVALID_PARAMETER` when any of these are not set as the Windows one does in [memory.c](https://github.com/rapid7/metasploit-payloads/blob/6bac251842e0140b78e6bd7e23e56688b83f34cb/c/meterpreter/source/extensions/stdapi/server/sys/process/memory.c#L98-L105).

The Windows one doesn't appear to allow a value of 0, so you can't use this to write to the NULL page, even if it were allocated. For the sake of simplicity, the Python one should also probably require a non-zero number.

For example, check with `if not (handle and base and size)` and not `if not (handle is not None and base is not None...)`.

**Comment by jmartin-r7** [comment]
@msjenkins-r7 test this please.

**Comment by smcintyre-r7** [comment]
@msjenkins-r7 test this please.

**Review by smcintyre-r7 — APPROVED** [review]
I tested this with Python 2.7 and 3.10 on a Server 2019 host. Everything worked as intended, but I pushed up one small change to use `None` for missing values instead of 0. As soon as the tests pass again, I'll get this landed. Thanks!

**Pull request merged**